### PR TITLE
Add checks in term and terms queries that input terms are not too long

### LINKED ISSUE #99802: Limit term query values to IndexWriter.MAX_TERM_LENGTH
### Description

There is a hard limit on the size of terms that can be indexed in lucene.  However, we don't check this at query time, which can lead to memory issues if the term query in question is case insensitive.  Very long terms are normally a sign of user error, so we should check up-front the length of the input and throw and exception if it is greater than `IndexWriter.MAX_TERM_LENGTH`.

### PR DESCRIPTION
**Pull request by romseygeek**
Lucene indexes do not allow terms of greater than 32k bytes long.  Any queries that
contain terms that exceed this length will by definition not match anything, and can
cause cluster instability by consuming large amounts of heap.  They are also generally
always a user error (for example, a termsquery that concatenates all its inputs into
a single string rather than splitting them into json arrays).

This commit adds some checking to Term and Terms query builders that will throw an
exception if any of their input terms are greater than the maximum allowed length by
the lucene IndexWriter.

Fixes #99802

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java [hunk 1]
```diff
@@ -211,9 +211,9 @@ public final int hashCode() {
      */
     static Object maybeConvertToBytesRef(Object obj) {
         if (obj instanceof String) {
-            return BytesRefs.toBytesRef(obj);
+            return BytesRefs.checkIndexableLength(BytesRefs.toBytesRef(obj));
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java:214]: Do you think it is worth checking the length even before converting String to BytesRef? Or it makes the code too complex?

**Review by mayya-sharipova — APPROVED** [review]
Thanks @romseygeek for a quick fix, LGTM!

File: server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java (hunk 1 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java:214]: There are complexities in converting between java String lengths and UTF8 encoded bytes which makes me think this is more failsafe.

**Review by martijnvg — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/common/lucene/BytesRefs.java [hunk 2]
```diff
@@ -39,4 +40,34 @@ public static BytesRef toBytesRef(Object value) {
...
+        return input;
+    }
+
+    /**
+     * Produces a UTF-string prefix of the input BytesRef.  If the prefix cutoff would produce
+     * ill-formed UTF, it falls back to the hexadecimal representation.
+     * @param input an input BytesRef
+     * @return a String prefix
+     */
+    public static String safeStringPrefix(BytesRef input, int prefixLength) {
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/common/lucene/BytesRefs.java:64]: Yep, that's a good call, will make this private.

**Comment by martijnvg** [comment]
Should this change be back ported to the 8.10 branch?

**Comment by romseygeek** [comment]
> Should this change be back ported to the 8.10 branch?

Yes, good idea.

**Pull request merged**